batch/batch/front_end/front_end.py

```diff
@@ -428,22 +455,37 @@ async def get_completed_batches_ordered_by_completed_time(request, userdata):
     return web.json_response(body)
 
 
-async def _get_jobs(
-    request: web.Request, batch_id: int, version: int, q: str, last_job_id: Optional[int]
+async def _get_job_group_jobs(
```

Comment:
Similar as before, keeping 'routes.get('/api/v1alpha/batches/completed')'
and renaming '_get_jobs' to '_get_job_group_jobs'.
Diff output:
```
+<<<<<<< HEAD
+@routes.get('/api/v1alpha/batches/completed')
+@auth.authenticated_users_only()
+async def get_completed_batches_ordered_by_completed_time(request, userdata):
+    db = request.app['db']
+    where_args = [userdata['username']]
+    wheres = [
+        'billing_project_users.`user` = %s',
+        'billing_project_users.billing_project = batches.billing_project',
+        'time_completed IS NOT NULL',
+        'NOT deleted',
+    ]
+
+    limit = 100
+    query_limit: str = request.query.get('limit')
+    if query_limit:
+        try:
+            limit = int(query_limit)
+        except ValueError as e:
+            raise web.HTTPBadRequest(reason=f'Bad value for "limit": {e}')
+
+    last_completed_timestamp = request.query.get('last_completed_timestamp')
+    if last_completed_timestamp:
+        where_args.append(int(last_completed_timestamp))
+        wheres.append('time_completed < %s')
+
+    sql = f"""
+SELECT batches.*,
+  job_groups_cancelled.id IS NOT NULL AS cancelled,
+  job_groups_n_jobs_in_complete_states.n_completed,
+  job_groups_n_jobs_in_complete_states.n_succeeded,
+  job_groups_n_jobs_in_complete_states.n_failed,
+  job_groups_n_jobs_in_complete_states.n_cancelled,
+  cost_t.*
+FROM batches
+LEFT JOIN billing_projects
+    ON batches.billing_project = billing_projects.name
+LEFT JOIN job_groups_n_jobs_in_complete_states
+    ON batches.id = job_groups_n_jobs_in_complete_states.id
+LEFT JOIN job_groups_cancelled
+    ON batches.id = job_groups_cancelled.id
+STRAIGHT_JOIN billing_project_users
+    ON batches.billing_project = billing_project_users.billing_project
+LEFT JOIN LATERAL (
+    SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
+    FROM (
+      SELECT batch_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+      FROM aggregated_job_group_resources_v3
+      WHERE batches.id = aggregated_job_group_resources_v3.batch_id
+      GROUP BY batch_id, resource_id
+    ) AS usage_t
+    LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
+    GROUP BY batch_id
+  ) AS cost_t ON TRUE
+WHERE
+    {' AND '.join(wheres)}
+ORDER BY time_completed DESC
+LIMIT %s;
+    """
+
+    records = [
+        batch async for batch in db.select_and_fetchall(sql, (*where_args, limit), query_name='get_completed_batches')
+    ]
+    # this comes out as a timestamp (rather than a formed date)
+    last_completed_timestamp = records[-1]['time_completed']
+    batches = [batch_record_to_dict(batch) for batch in records]
+    body = {'batches': batches}
+    if len(batches) == limit:
+        body['last_completed_timestamp'] = last_completed_timestamp
+    return web.json_response(body)
+
+
+async def _get_jobs(
+    request: web.Request, batch_id: int, version: int, q: str, last_job_id: Optional[int]
+||||||| dc7fce0e3
+async def _get_jobs(
+    request: web.Request, batch_id: int, version: int, q: str, last_job_id: Optional[int]
+=======
 async def _get_job_group_jobs(
     request: web.Request,
     batch_id: int,
@@ -303,6 +478,7 @@ async def _get_job_group_jobs(
     q: str,
     last_job_id: Optional[int],
     recursive: bool,
+>>>>>>> 13de4e6ff
 ) -> GetJobsResponseV1Alpha:
     db = request.app['db']
 
@@ -373,6 +549,108 @@ async def _api_get_job_group_jobs(request, batch_id: int, job_group_id: int, ver
     return json_response(resp)
```

Reply:
Yes. Though the same observation about checking whether this needs changes for job groups applies too…

Reply:
I can see the sql in our route '/api/v1alpha/batches/completed' was adjusted copy of what was in _get_batch original versions prior to job_groups change. I guess we would need to bring over that change.
I have noticed ROOT_JOB_GROUP_ID is used in API points where job_group is not available, e.g.:
@routes.get('/api/v1alpha/batches/{batch_id}/jobs')
Let me try to adjust it.

Reply:
I did updated the sql using the old and new versions of sql in get_batch function.